zebra-consensus/src/verify.rs

```diff
@@ -0,0 +1,76 @@
+//! Block verification and chain state updates for Zebra.
+//!
+//! Verification occurs in multiple stages:
+//!   - getting blocks (disk- or network-bound)
+//!   - context-free verification of signatures, proofs, and scripts (CPU-bound)
+//!   - context-dependent verification of the chain state (awaits a verified parent block)
+//!
+//! Verification is provided via a `tower::Service`, to support backpressure and batch
+//! verification.
+
+use std::{
+    error::Error,
+    future::Future,
+    pin::Pin,
+    task::{Context, Poll},
+};
+use tower::{buffer::Buffer, Service};
+
+use zebra_chain::block::{Block, BlockHeaderHash};
+
+mod script;
+mod transaction;
+
+/// Block verification service.
+///
+/// After verification, blocks and their associated transactions are added to
+/// `zebra_state::ZebraState`.
+#[derive(Default)]
+struct BlockVerifier {}
+
+/// The result type for the BlockVerifier Service.
+type Response = BlockHeaderHash;
+
+/// The error type for the BlockVerifier Service.
+// TODO(jlusby): Error = Report ?
+type ServiceError = Box<dyn Error + Send + Sync + 'static>;
+
+/// The BlockVerifier service implementation.
+impl Service<Block> for BlockVerifier {
+    type Response = Response;
+    type Error = ServiceError;
+    type Future =
+        Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send + 'static>>;
+
+    fn poll_ready(&mut self, context: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        // TODO(teor): is this a shared state?
+        let mut state_service = zebra_state::in_memory::init();
+        state_service.poll_ready(context)
+    }
+
+    fn call(&mut self, block: Block) -> Self::Future {
+        let mut state_service = zebra_state::in_memory::init();
```

Comment:
This should also just reference the service in `self`

Reply:
Done in d19fd2ec504aa4fc813338ae4e0b5f1dbb169294 and previous commits.